### PR TITLE
`ruff` fixes

### PR DESCRIPTION
**Pull request by DanielYang59**
### Summary

- `test_wandb_log_frequency` use temporary path to avoid creating files locally

### Relocate to another PR

- Build against NumPy 2.0, https://numpy.org/doc/stable/dev/depending_on_numpy.html
- Ruff `NPY201` rule run seems to pass (https://numpy.org/doc/stable/numpy_2_0_migration_guide.html):
    > Many of the changes covered in the 2.0 release notes and in this migration guide can be automatically adapted in downstream code with a dedicated [Ruff](https://docs.astral.sh/ruff/) rule, namely rule [NPY201](https://docs.astral.sh/ruff/rules/numpy2-deprecation/).

### CODE REVIEW

**Comment by DanielYang59** [comment]
@janosh Can you please approve the workflow? Thanks!

**Comment by DanielYang59** [comment]
Thanks! I believe the test failure is owing to the lack of support of NumPy 2.0 from `pymatgen`'s side https://github.com/materialsproject/pymatgen/issues/3953. Not sure what's the best practice at this moment as `pymatgen` is waiting for `chgnet` to support NumPy 2.0 too https://github.com/materialsproject/pymatgen/pull/3894#issuecomment-2190098821

**Comment by janosh** [comment]
Pymatgen should go first and ignore failing chgnet tests

**Comment by DanielYang59** [comment]
@janosh It looks like you have to approve the workflow again for some reason?

**Comment by janosh** [comment]
until your first PR to a repo is merged, workflows need to be re-approved on every commit. it's quite annoying

**Comment by DanielYang59** [comment]
Hi @janosh if i remember correctly, this cannot be merged until a new release of `pymatgen` is issued.

**Comment by janosh** [comment]
@DanielYang59 thanks for the reminder. [pymatgen v2024.9.10](https://github.com/materialsproject/pymatgen/releases/tag/v2024.9.10) should be on PyPI in a few minutes

**Comment by DanielYang59** [comment]
@janosh Have to ping you to approve it (again), hopefully no test fails this time

File: chgnet/model/composition_model.py [hunk 1]
```diff
@@ -146,7 +146,7 @@ def fit(
             if isinstance(structure, Structure):
                 atomic_number = torch.tensor(
                     [site.specie.Z for site in structure],
-                    dtype=int,
+                    dtype=np.int64,
```
Comment (DanielYang59) [inline chgnet/model/composition_model.py:149]: I'm a bit unsure about pytorch

**Comment by DanielYang59** [comment]
@janosh Should be good this time :)

File: tests/test_crystal_graph.py [hunk 2]
```diff
@@ -127,55 +126,57 @@ def test_crystal_graph_different_cutoff_fast():
 
 
 def test_crystal_graph_perturb_legacy():
-    np.random.seed(0)
```
[the remark this answers is not in the document]
Comment (janosh) [reply tests/test_crystal_graph.py:130]: thank you! your changes in b4f05c5 look good 👍

[the remark this answers is not in the document]
Comment (DanielYang59) [reply tests/test_crystal_graph.py:130]: No problem, also fixed d0c6b5b99f577fcb92304c94c48f213b674a6c77 from 73d6096790bbb1e20d5f2e1e23119cca34e54483

**Comment by DanielYang59** [comment]
Looks like [something is still not right with int type for windows](https://github.com/CederGroupHub/chgnet/actions/runs/10811604391/job/29992113246?pr=184) (with pymatgen 2024.9.10), changing to my Windows PC for debugging...


```
____________________ ERROR collecting tests/test_model.py _____________________
tests\test_model.py:14: in <module>
    graph = CrystalGraphConverter()(structure, graph_id="test-model")
C:\hostedtoolcache\windows\Python\3.12.5\x64\Lib\site-packages\torch\nn\modules\module.py:1511: in _wrapped_call_impl
    return self._call_impl(*args, **kwargs)
C:\hostedtoolcache\windows\Python\3.12.5\x64\Lib\site-packages\torch\nn\modules\module.py:1520: in _call_impl
    return forward_call(*args, **kwargs)
chgnet\graph\converter.py:135: in forward
    graph = self.create_graph(
chgnet\graph\converter.py:255: in _create_graph_fast
    nodes, dir_edges_list, undir_edges_list, undirected_edges = make_graph(
chgnet\graph\cygraph.pyx:65: in chgnet.graph.cygraph.make_graph
    const long[::1] center_index,
E   ValueError: Buffer dtype mismatch, expected 'const long' but got 'long long'
```

**Comment by janosh** [comment]
based on the [numpy docs](https://numpy.org/doc/stable/user/basics.types.html#relationship-between-numpy-data-types-and-c-data-types) for `intp_t`, maybe we should change the type signature to this?

```pyx
def make_graph(
    const intp_t[::1] center_index,
    const intp_t n_e,
    const intp_t[::1] neighbor_index,
    const intp_t[:, ::1] image,
    const float64_t[::1] distance,
    const intp_t num_atoms
):
```

prev

```pyx
def make_graph(
    const long[::1] center_index,
    const long n_e,
    const long[::1] neighbor_index,
    const long[:, ::1] image,
    const double[::1] distance,
    const long num_atoms
):
```

**Comment by DanielYang59** [comment]
Yes! Changed to numpy/cython type, tests seem to pass locally on my Windows PC

File: pyproject.toml
Comment (janosh) [inline pyproject.toml:1]: lol, best commit message i've read in weeks 🤣

Comment (DanielYang59) [reply pyproject.toml:1]: :) My bad, looks like my Windows PC is indeed very slow not only for install but search, I search those ruff rules and remove them if I found no match, turns out it's my PC being too slow to find a match 😄 

Would fix those.

But the build failure is a bit unexpected, I thought `numpy` is already included for build, need a closer look :(:

```
[build-system]
requires = [
    "Cython",
    "setuptools>=65",
    "wheel",
    "numpy>=2.0.0",
]
build-backend = "setuptools.build_meta"
```

```
Traceback (most recent call last):
  File "D:\a\chgnet\chgnet\setup.py", line 3, in <module>
    import numpy as np
ModuleNotFoundError: No module named 'numpy'
```

I wish this terrible approve every commit strategy could be changed soon.

Comment (janosh) [reply pyproject.toml:1]: once you have your first PR merged into `chgnet`, the need to approve each commit for CI will go away for this repo

File: .github/workflows/test.yml
Comment (janosh) [inline .github/workflows/test.yml:1]: good call, that's what i wanted to try next as well but couldn't force push over my previous "try" commit

Comment (DanielYang59) [reply .github/workflows/test.yml:1]: I was being stupid, I could have just [ran workflow on my fork](https://github.com/DanielYang59/chgnet/actions/runs/10814629343/job/30001543941)

File: chgnet/graph/converter.py [hunk 3]
```diff
@@ -150,7 +150,7 @@ def forward(
             # Report structures that failed creating bond graph
             # This happen occasionally with pymatgen version issue
             structure.to(filename="bond_graph_error.cif")
-            raise SystemExit(
```
Comment (DanielYang59) [inline chgnet/graph/converter.py:153]: Not sure if this is intended, but I don't think `SystemExit` should be raised here?

**Comment by DanielYang59** [comment]
Looks like there is another error [just on windows for some reason](https://github.com/DanielYang59/chgnet/actions/runs/10815019610/job/30002865127):

```
____________________ ERROR collecting tests/test_model.py _____________________
chgnet\graph\converter.py:146: in forward
    bond_graph, undirected2directed = graph.line_graph_adjacency_list(
chgnet\graph\graph.py:272: in line_graph_adjacency_list
    raise ValueError(
E   ValueError: Error: number of directed edges=672 != 2 * number of undirected edges=473!This indicates directed edges are not complete
The above exception was the direct cause of the following exception:
tests\test_model.py:14: in <module>
    graph = CrystalGraphConverter()(structure, graph_id="test-model")
C:\hostedtoolcache\windows\Python\3.10.11\x64\lib\site-packages\torch\nn\modules\module.py:1553: in _wrapped_call_impl
    return self._call_impl(*args, **kwargs)
C:\hostedtoolcache\windows\Python\3.10.11\x64\lib\site-packages\torch\nn\modules\module.py:1562: in _call_impl
    return forward_call(*args, **kwargs)
chgnet\graph\converter.py:153: in forward
    raise RuntimeError(
E   RuntimeError: Failed creating bond graph for test-model, check bond_graph_error.cif
```

Would have a look tomorrow.

**Comment by DanielYang59** [comment]
Also there is another [seemingly error for Ubuntu/MacOS](https://github.com/DanielYang59/chgnet/actions/runs/10815019610/job/30002863491), not sure if this is related to this PR:


```
tests/test_encoders.py .................
!!!!!! the two directed edges are equal but this operation is not supposed to happen
!!!!!! the two directed edges are equal but this operation is not supposed to happen
tests/test_graph.py ........
```

**Comment by janosh** [comment]
maybe @BowenD-UCB can chime on this test failure which oddly only happens on Windows? perhaps this signals that one of the `dtype` changes in `make_graph` is still platform dependent or needs to be reverted?

```py
chgnet\graph\converter.py:148: in forward
    bond_graph, undirected2directed = graph.line_graph_adjacency_list(
chgnet\graph\graph.py:272: in line_graph_adjacency_list
    raise ValueError(
E   ValueError: Error: number of directed edges=672 != 2 * number of undirected edges=473!This indicates directed edges are not complete
```

**Comment by BowenD-UCB** [comment]
> maybe @BowenD-UCB can chime on this test failure which oddly only happens on Windows? perhaps this signals that one of the `dtype` changes in `make_graph` is still platform dependent or needs to be reverted?
> 
> ```python
> chgnet\graph\converter.py:148: in forward
>     bond_graph, undirected2directed = graph.line_graph_adjacency_list(
> chgnet\graph\graph.py:272: in line_graph_adjacency_list
>     raise ValueError(
> E   ValueError: Error: number of directed edges=672 != 2 * number of undirected edges=473!This indicates directed edges are not complete
> ```

@DanielYang59 @janosh 
Yes I'll look into this later.
Sorry about the delay on the PR, been a bit busy lately.

**Comment by DanielYang59** [comment]
> perhaps this signals that one of the dtype changes in make_graph is still platform dependent or needs to be reverted?

I think so, we should focus on inferred types, i.e. `np.array([0, 1])` whose type would still be platform dependent. 

Instead of just changing the type in tests, we should also update the code where `int` is expected and automatically cast the type to int64. In this regard, I'm not familiar with the code base so if Bowen could give a hand that would very helpful.

>  Sorry about the delay on the PR, been a bit busy lately.

It has not been delayed at all :) `pymatgen` just unblocked this.

**Comment by DanielYang59** [comment]
@janosh I reverted all (I believe) numpy related change from this PR, and relocate NumPy v2 migration to another PR as it turns out there's more debugging needed than I expected.

File: chgnet/trainer/trainer.py [hunk 4]
```diff
@@ -285,7 +285,7 @@ def train(
             raise ValueError("Model needs to be initialized")
         global best_checkpoint  # noqa: PLW0603
         if save_dir is None:
-            save_dir = f"{datetime.now():%m-%d-%Y}"
+            save_dir = f"{datetime.datetime.now(tz=datetime.timezone.utc):%m-%d-%Y}"
```
Comment (janosh) [inline chgnet/trainer/trainer.py:288]: maybe better to use the user's own timezone here? not sure? would be
```suggestion
            save_dir = f"{datetime.datetime.now().astimezone():%m-%d-%Y}"
```

File: chgnet/model/dynamics.py [hunk 5]
```diff
@@ -294,7 +295,7 @@ def relax(
                 cry_obs = CrystalFeasObserver(atoms)
 
             if relax_cell:
-                atoms = ase_filter(atoms)
+                atoms = FrechetCellFilter(atoms)
```
Comment (janosh) [inline chgnet/model/dynamics.py:298]: i think this should stay `ase_filter`

File: tests/test_crystal_graph.py [hunk 6]
```diff
@@ -127,55 +126,57 @@ def test_crystal_graph_different_cutoff_fast():
...
     structure_perturbed = structure.copy()
-    structure_perturbed.perturb(distance=0.1)
+    fixed_rng = np.random.default_rng(0)
+    with patch("numpy.random.default_rng", return_value=fixed_rng):
+        structure_perturbed.perturb(distance=0.1)
```
Comment (janosh) [inline tests/test_crystal_graph.py:132]: we should probably add a `seed` kwarg to `Structure.perturb` that's simply passed into `np.random.default_rng(seed=seed)` over in `pymatgen`

File: chgnet/graph/graph.py [hunk 7]
```diff
@@ -271,7 +271,7 @@ def line_graph_adjacency_list(self, cutoff) -> tuple[list[list[int]], list[int]]
         if len(self.directed_edges_list) != 2 * len(self.undirected_edges_list):
             raise ValueError(
                 f"Error: number of directed edges={len(self.directed_edges_list)} != 2 "
-                f"* number of undirected edges={len(self.directed_edges_list)}!"
+                f"* number of undirected edges={len(self.undirected_edges_list)}!"
```
Comment (janosh) [inline chgnet/graph/graph.py:274]: good catch! 👍

**Review by janosh — CHANGES_REQUESTED** [review]

File: chgnet/model/dynamics.py (hunk 5 above)
Comment (DanielYang59) [reply chgnet/model/dynamics.py:298]: So we just ignore this following warning?

```
/Users/runner/work/chgnet/chgnet/chgnet/model/dynamics.py:297: DeprecationWarning: Use FrechetCellFilter for better convergence w.r.t. cell variables.

    atoms = ase_filter(atoms)
```

Comment (janosh) [reply chgnet/model/dynamics.py:298]: `FrechetCellFilter` should be the default but there needs to be a way to customize it for people who want to compare with legacy behavior from last year when `FrechetCellFilter` didn't exist yet and `ExpCellFilter` was the default

Comment (DanielYang59) [reply chgnet/model/dynamics.py]: Sure reverted in aeb90530c51d80708031b66248430fc33d330c39

File: chgnet/graph/graph.py (hunk 7 above)
Comment (DanielYang59) [reply chgnet/graph/graph.py:274]: No problem, wouldn't have noticed it if not for the test message like `672 != 672` LOL

File: chgnet/trainer/trainer.py (hunk 4 above)
Comment (janosh) [reply chgnet/trainer/trainer.py:288]: good points. using UTC means you often have to do the mental math from your timezone to UTC to understand how long ago a file/dir was actually written but that's probably the lesser of two evils

File: tests/test_crystal_graph.py (hunk 6 above)
Comment (DanielYang59) [reply tests/test_crystal_graph.py:132]: That's a good point, especially with the generator implementation. 

I didn't realize this downside before but it turns out the new generator implementation every `rng` is a isolated instance, making it impossible to set a global seed and control all random states (or it's just me unaware of that).

**Review by janosh — COMMENTED** [review]
thanks @DanielYang59! 👍

**Comment by DanielYang59** [comment]
No problem, thanks for merging this.

**Pull request merged**